lib/app/docker/scanningimageservice.go

```diff
@@ -0,0 +1,145 @@
+/*
+Copyright 2020 Gravitational, Inc.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package docker
+
+import (
+	"fmt"
+	"strings"
+
+	context "github.com/docker/distribution/context"
+	registryauth "github.com/docker/distribution/registry/client/auth"
+	"github.com/gravitational/gravity/lib/utils"
+	"github.com/gravitational/trace"
+)
+
+// ScanConfig represents configuration used to push images to a docker repository configured for scanning those
+// images for vulnerabilities or other problems
+type ScanConfig struct {
+	// Remote Repository is the docker repository to use to scan the images.
+	// Example: gravitational/gravity-scan
+	RemoteRepository string
+	// TagPrefix is a string to prepend to each images tag, in order to help identify the source of the image.
+	// Example: 7.0.0 for a release of gravity 7.0.0
+	TagPrefix string
+}
+
+// NewScanningImageService creates an image service that rewrites image paths to a
+// repository used for scanning those images.
+func NewScanningImageService(req RegistryConnectionRequest, conf ScanConfig) (ImageService, error) {
+	base, err := NewImageService(req)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &scanningImageService{
+		imageService: base.(*imageService),
```

Comment:
To avoid the type cast, `NewImageService` could return an exported type instead of the interface (optionally, with all fields unexported if it needs to continue being opaque to the caller).